src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs

```diff
@@ -190,19 +200,25 @@ private void DisconnectFromAllTaggers()
             private void RegisterNotification(Action action)
             {
                 var token = _owner._listener.BeginAsyncOperation(GetType().Name + "RegisterNotification");
-                _owner._notificationService.RegisterNotification(action, token);
+                _owner._notificationService.RegisterNotification(action, token, _shutdownCancellationSource.Token);
             }
 
             private void OnDiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs e)
             {
                 lock (_taskGate)
                 {
+                    if (e.DocumentId != _currentDocumentId)
+                    {
+                        // ignore diagnostics that belong to different documents
```

Comment:
is this racey?  what happens if our buffer now maps to a different document?  will we suddenly ignore events about the diagnostics for the buffer before the document changed?

Reply:
no. it won't. if it does, then that is an issue in Workspace.Events it is listening to.

Reply:
if there is an issue in Workspace events, and if we are trying to fix that by moving everything on UI thread and do filtering there, that is exactly trying to fixing the symptom not root cause of it.